File: backend/services/egress_guard.js
'use strict'

/**
 * Egress guard — the never-leak gate for the compute fabric.
 *
 * NOTHING leaves this machine to a remote destination (paired peer, rented GPU,
 * or external API) without passing through here first. CLAUDE.md #5/#20: secrets
 * must never be sent off-box; sensitive data must be redacted; data classification
 * rules are enforced server-side, deny-by-default, fail-closed.
 *
 *   classify(payload)            → highest data classification found
 *                                  (public < internal < pii < secret)
 *   guard(payload, destTier)     → { action, payload, classification, ... }
 *                                  action ∈ allow | redact | block
 *   isEndpointAllowed(url)       → only private-LAN / https endpoints dispatch
 *
 * Redaction reuses backend/services/forge_learning.scrubSecretsFromLearningData
 * (battle-tested deep secret scrub). Policy: runtime/config/egress_policy.json,
 * with baked-in defaults so the gate never crashes open.
 *
 * Pure, no network, never throws. On ANY error it returns BLOCK (fail-closed).
 */

const fs = require('fs')
const path = require('path')
const { scrubSecretsFromLearningData } = require('./forge_learning')

const DEFAULTS = {
  destination_tiers: { local: 0, peer_trusted: 1, rented_trusted: 2, external_api: 3 },
  classification_rank: { public: 0, internal: 1, pii: 2, secret: 3 },
  policy_matrix: {
    local: { secret: 'allow', pii: 'allow', internal: 'allow', public: 'allow' },
    peer_trusted: { secret: 'block', pii: 'redact', internal: 'allow', public: 'allow' },
    rented_trusted: { secret: 'block', pii: 'redact', internal: 'redact', public: 'allow' },
    external_api: { secret: 'block', pii: 'redact', internal: 'redact', public: 'allow' },
  },
  caps: { max_payload_bytes: 2097152, dispatch_timeout_ms: 60000, max_result_bytes: 8388608 },
  endpoint_allow: {
    url_patterns: [
      '^https://',
      '^http://(localhost|127\\.0\\.0\\.1)(:\\d+)?(/|$)',
      '^http://10\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}(:\\d+)?(/|$)',
      '^http://192\\.168\\.\\d{1,3}\\.\\d{1,3}(:\\d+)?(/|$)',
      '^http://172\\.(1[6-9]|2\\d|3[01])\\.\\d{1,3}\\.\\d{1,3}(:\\d+)?(/|$)',
    ],
  },
}

const CONFIG_PATH = path.join(__dirname, '..', '..', 'runtime', 'config', 'egress_policy.json')

let _cfg = null
function loadPolicy() {
  if (_cfg) return _cfg
  let f = {}
  try { f = JSON.parse(fs.readFileSync(CONFIG_PATH, 'utf8')) } catch (_) { f = {} }
  _cfg = {
    ...DEFAULTS, ...f,
    destination_tiers: { ...DEFAULTS.destination_tiers, ...(f.destination_tiers || {}) },
    classification_rank: { ...DEFAULTS.classification_rank, ...(f.classification_rank || {}) },
    policy_matrix: { ...DEFAULTS.policy_matrix, ...(f.policy_matrix || {}) },
    caps: { ...DEFAULTS.caps, ...(f.caps || {}) },
    endpoint_allow: { ...DEFAULTS.endpoint_allow, ...(f.endpoint_allow || {}) },
  }
  return _cfg
}
function _resetPolicy() { _cfg = null }

// PII / internal detectors (secret detection is delegated to the scrubber).
const PII_RE = [
  /[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/,                 // email
  /\b(?:\+?\d[\d\s().-]{7,}\d)\b/,                                   // phone-ish
  /\b(?:\d[ -]*?){13,19}\b/,                                        // card-ish
]
const INTERNAL_RE = [
  /(?:^|[\s"'`(=])\/(?:home|root|etc|var|usr|opt|Users)\//,         // absolute fs paths
  /\b[A-Za-z0-9_-]+\.(?:local|internal|lan|home)\b/,                // internal hostnames
]

const _stringify = (p) => {
  try { return typeof p === 'string' ? p : JSON.stringify(p) } catch (_) { return String(p) }
}

// Did the deep secret-scrub change anything? If so, the payload contained a secret.
function containsSecret(payload) {
  try {
    const before = _stringify(payload)
    const after = _stringify(scrubSecretsFromLearningData(payload))
    return before !== after
  } catch (_) { return true } // fail-closed: assume secret on error
}

function classify(payload) {
  try {
    if (containsSecret(payload)) return 'secret'
    const text = _stringify(payload)
    if (PII_RE.some(re => re.test(text))) return 'pii'
    if (INTERNAL_RE.some(re => re.test(text))) return 'internal'
    return 'public'
  } catch (_) { return 'secret' } // fail-closed
}

// Redact PII spans (email/phone/card) in a single string.
function redactPIIString(s) {
  let out = String(s)
  for (const re of PII_RE) out = out.replace(new RegExp(re.source, 'g'), '[REDACTED_PII]')
  return out
}

// Walk a JSON-ish structure applying fn to every string value. Bounded depth so a
// hostile deeply-nested payload can't blow the stack.
function _deepMapStrings(v, fn, depth = 0) {
  if (depth > 64) return '[TRUNCATED_DEPTH]'
  if (typeof v === 'string') return fn(v)
  if (Array.isArray(v)) return v.map(x => _deepMapStrings(x, fn, depth + 1))
  if (v && typeof v === 'object') {
    const out = {}
    for (const [k, val] of Object.entries(v)) {
      if (k === '__proto__' || k === 'constructor' || k === 'prototype') continue // strip pollution keys
      out[k] = _deepMapStrings(val, fn, depth + 1)
    }
    return out
  }
  return v
}

// Full redaction: secrets (deep scrub + key-drop) AND PII. Used whenever the
// policy action is 'redact'.
function redact(payload) {
  try {
    const noSecrets = scrubSecretsFromLearningData(payload)
    return _deepMapStrings(noSecrets, redactPIIString)
  } catch (_) { return '[REDACTED]' }
}

function isEndpointAllowed(url) {
  try {
    const u = String(url || '')
    if (!u) return false
    return (loadPolicy().endpoint_allow.url_patterns || []).some(p => {
      try { return new RegExp(p).test(u) } catch (_) { return false }
    })
  } catch (_) { return false }
}

function payloadBytes(payload) {
  try { return Buffer.byteLength(_stringify(payload), 'utf8') } catch (_) { return Infinity }
}

/**
 * Decide whether `payload` may leave to `destinationTier`. NEVER throws.
 * Returns { action, payload, classification, tier, reason }.
 *   - action 'block'  → do not send. payload is null.
 *   - action 'redact' → send the returned (scrubbed) payload.
 *   - action 'allow'  → send payload as-is.
 */
function guard(payload, destinationTier) {
  try {
    const cfg = loadPolicy()
    const tier = String(destinationTier || '')
    // Unknown destination → deny-by-default.
    if (!(tier in cfg.policy_matrix)) {
      return { action: 'block', payload: null, classification: 'unknown', tier, reason: `unknown destination tier '${tier}'` }
    }
    // Oversize → block (an attacker can't exfiltrate a huge blob).
    const bytes = payloadBytes(payload)
    if (bytes > Number(cfg.caps.max_payload_bytes)) {
      return { action: 'block', payload: null, classification: 'oversize', tier, reason: `payload ${bytes}B exceeds cap ${cfg.caps.max_payload_bytes}B` }
    }
    const cls = classify(payload)
    const action = (cfg.policy_matrix[tier] && cfg.policy_matrix[tier][cls]) || 'block' // missing rule → block
    if (action === 'allow') return { action: 'allow', payload, classification: cls, tier, reason: 'within policy' }
    if (action === 'redact') return { action: 'redact', payload: redact(payload), classification: cls, tier, reason: `${cls} redacted for ${tier}` }
    return { action: 'block', payload: null, classification: cls, tier, reason: `${cls} not permitted to ${tier}` }
  } catch (e) {
    return { action: 'block', payload: null, classification: 'error', tier: String(destinationTier || ''), reason: 'egress guard error (fail-closed)' }
  }
}

// Structurally CONTAIN an untrusted inbound value: a compromised/malware-infected
// worker must not be able to contaminate this process. We accept ONLY plain JSON
// primitives, drop prototype-pollution keys (__proto__/constructor/prototype),
// bound depth + string length, and reject functions/symbols/other live objects.
// The result is therefore inert data — it can never be a code/object payload.
function containValue(v, depth = 0) {
  if (depth > 64) return '[TRUNCATED_DEPTH]'
  const t = typeof v
  if (v === null || t === 'boolean' || t === 'number') return v
  if (t === 'string') return v.length > 200000 ? v.slice(0, 200000) + '…[TRUNCATED]' : v
  if (Array.isArray(v)) return v.slice(0, 10000).map(x => containValue(x, depth + 1))
  if (t === 'object') {
    const out = {}
    for (const [k, val] of Object.entries(v)) {
      if (k === '__proto__' || k === 'constructor' || k === 'prototype') continue
      out[String(k).slice(0, 200)] = containValue(val, depth + 1)
    }
    return out
  }
  return undefined // functions, symbols, bigint, undefined → dropped entirely
}

// Inbound results from a worker are UNTRUSTED (CLAUDE.md #1/#10: agent outputs are
// hostile). Containment pipeline: size-cap → structural contain (anti-malware) →
// secret+PII redact. The returned value is inert data, never code, and is tagged
// so downstream code can never silently trust or execute it.
function scanResult(result) {
  try {
    const cfg = loadPolicy()
    const bytes = payloadBytes(result)
    if (bytes > Number(cfg.caps.max_result_bytes)) {
      return { ok: false, result: null, reason: `result ${bytes}B exceeds cap ${cfg.caps.max_result_bytes}B` }
    }
    const had_secret = containsSecret(result)
    const contained = containValue(result)
    const safe = redact(contained)
    return { ok: true, result: safe, had_secret, _untrusted: true }
  } catch (_) { return { ok: false, result: null, reason: 'result scan error (fail-closed)' } }
}

module.exports = {
  loadPolicy, classify, redact, redactPIIString, guard, isEndpointAllowed,
  scanResult, containValue, containsSecret, payloadBytes, _resetPolicy,
}

File: runtime/core/egress_guard.py
"""Egress guard (Python) — never-leak gate for the LLM provider path.

Mirrors backend/services/egress_guard.js and shares runtime/config/egress_policy.json.
Before any prompt/message leaves this machine to an EXTERNAL provider (Anthropic /
OpenAI / OpenRouter) it must pass through here: secrets are BLOCKED off-box, PII is
redacted, and the destination policy is enforced deny-by-default + fail-closed.

CLAUDE.md #5/#20: do not send sensitive local data to external models; redact prompts
before remote fallback; prevent silent leakage.

Reuses runtime/core/output_guard.redact_pii_echo for PII. Pure, never raises
(on any error → BLOCK).
"""

from __future__ import annotations

import json
import logging
import os
import re
from pathlib import Path
from typing import Any

logger = logging.getLogger(__name__)

_CONFIG_PATH = Path(__file__).resolve().parents[1] / "config" / "egress_policy.json"

_DEFAULTS: dict[str, Any] = {
    "destination_tiers": {"local": 0, "peer_trusted": 1, "rented_trusted": 2, "external_api": 3},
    "classification_rank": {"public": 0, "internal": 1, "pii": 2, "secret": 3},
    "policy_matrix": {
        "local": {"secret": "allow", "pii": "allow", "internal": "allow", "public": "allow"},
        "peer_trusted": {"secret": "block", "pii": "redact", "internal": "allow", "public": "allow"},
        "rented_trusted": {"secret": "block", "pii": "redact", "internal": "redact", "public": "allow"},
        "external_api": {"secret": "block", "pii": "redact", "internal": "redact", "public": "allow"},
    },
    "caps": {"max_payload_bytes": 2097152, "dispatch_timeout_ms": 60000, "max_result_bytes": 8388608},
}

# Secret shapes that must NEVER leave the box (mirror of the Node scrubber).
_SECRET_PATTERNS = [
    re.compile(r"sk-ant-[A-Za-z0-9_\-]{16,}"),
    re.compile(r"sk-[A-Za-z0-9]{20,}"),
    re.compile(r"gh[pousr]_[A-Za-z0-9]{20,}"),
    re.compile(r"eyJ[A-Za-z0-9_\-]+\.eyJ[A-Za-z0-9_\-]+\.[A-Za-z0-9_\-]+"),  # JWT
    re.compile(r"(?:AKIA|ASIA)[A-Z0-9]{16}"),
    re.compile(r"-----BEGIN [A-Z ]*PRIVATE KEY-----[\s\S]+?-----END [A-Z ]*PRIVATE KEY-----"),
    re.compile(r"(?im)^([A-Z][A-Z0-9_]+)=\S+$"),  # .env-style KEY=value
    re.compile(r"(?i)(?:api[_-]?key|secret|token|password|passwd|bearer)\s*[:=]\s*['\"]?[A-Za-z0-9+/=._\-]{12,}"),
]
_PII_PATTERNS = [
    re.compile(r"[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}"),  # email
    re.compile(r"\b(?:\+?\d[\d\s().\-]{7,}\d)\b"),                      # phone-ish (sync w/ JS mirror)
    re.compile(r"\b(?:\d[ \-]*?){13,19}\b"),                           # card-ish
]
_INTERNAL_PATTERNS = [
    re.compile(r"(?:^|[\s\"'`(=])/(?:home|root|etc|var|usr|opt|Users)/"),
    re.compile(r"\b[A-Za-z0-9_\-]+\.(?:local|internal|lan|home)\b"),
]

_cfg: dict[str, Any] | None = None


def _deep_merge(base: dict, override: dict) -> dict:
    out = dict(base)
    for k, v in (override or {}).items():
        out[k] = _deep_merge(out[k], v) if isinstance(v, dict) and isinstance(out.get(k), dict) else v
    return out


def load_policy() -> dict[str, Any]:
    global _cfg
    if _cfg is not None:
        return _cfg
    f: dict[str, Any] = {}
    try:
        f = json.loads(_CONFIG_PATH.read_text(encoding="utf-8"))
    except Exception as e:  # noqa: BLE001 — config problems must not break routing
        logger.debug("egress_guard: using defaults (%s)", e)
        f = {}
    _cfg = _deep_merge(_DEFAULTS, f if isinstance(f, dict) else {})
    return _cfg


def reload() -> None:
    global _cfg
    _cfg = None


def gate_enabled() -> bool:
    return os.environ.get("EGRESS_GUARD", "1") != "0"


def _text_of(payload: Any) -> str:
    if isinstance(payload, str):
        return payload
    try:
        return json.dumps(payload, default=str)
    except Exception:  # noqa: BLE001
        return str(payload)


def contains_secret(payload: Any) -> bool:
    try:
        t = _text_of(payload)
        return any(p.search(t) for p in _SECRET_PATTERNS)
    except Exception:  # noqa: BLE001
        return True  # fail-closed


def classify(payload: Any) -> str:
    try:
        if contains_secret(payload):
            return "secret"
        t = _text_of(payload)
        if any(p.search(t) for p in _PII_PATTERNS):
            return "pii"
        if any(p.search(t) for p in _INTERNAL_PATTERNS):
            return "internal"
        return "public"
    except Exception:  # noqa: BLE001
        return "secret"  # fail-closed


def _redact_text(s: str) -> str:
    out = s
    for p in _SECRET_PATTERNS:
        out = p.sub("[REDACTED]", out)
    # Reuse the project's PII redactor when available; fall back to local patterns.
    try:
        from core.output_guard import redact_pii_echo
        out = redact_pii_echo(out)
    except Exception:  # noqa: BLE001
        for p in _PII_PATTERNS:
            out = p.sub("[REDACTED_PII]", out)
    return out


def redact(payload: Any, _depth: int = 0) -> Any:
    if _depth > 64:
        return "[TRUNCATED_DEPTH]"
    try:
        if isinstance(payload, str):
            return _redact_text(payload)
        if isinstance(payload, list):
            return [redact(x, _depth + 1) for x in payload]
        if isinstance(payload, dict):
            out = {}
            for k, v in payload.items():
                if str(k) in ("__proto__", "constructor", "prototype"):
                    continue
                # Drop secret-named keys entirely.
                if re.fullmatch(r"(?i)password|passwd|secret|token|api_key|apikey|bearer|auth_token|private_key|access_key|session_id|cookie", str(k)):
                    out[k] = "[REDACTED]"
                else:
                    out[k] = redact(v, _depth + 1)
            return out
        return payload
    except Exception:  # noqa: BLE001
        return "[REDACTED]"


def guard(payload: Any, destination_tier: str) -> dict[str, Any]:
    """Decide whether `payload` may leave to `destination_tier`. Never raises.

    Returns {action, payload, classification, tier, reason}. action ∈ allow|redact|block.
    """
    try:
        if not gate_enabled():
            return {"action": "allow", "payload": payload, "classification": "ungated", "tier": destination_tier, "reason": "EGRESS_GUARD=0"}
        cfg = load_policy()
        tier = str(destination_tier or "")
        matrix = cfg["policy_matrix"]
        if tier not in matrix:
            return {"action": "block", "payload": None, "classification": "unknown", "tier": tier, "reason": f"unknown destination tier '{tier}'"}
        text = _text_of(payload)
        if len(text.encode("utf-8", "ignore")) > int(cfg["caps"]["max_payload_bytes"]):
            return {"action": "block", "payload": None, "classification": "oversize", "tier": tier, "reason": "payload exceeds size cap"}
        cls = classify(payload)
        action = matrix[tier].get(cls, "block")  # missing rule → block
        if action == "allow":
            return {"action": "allow", "payload": payload, "classification": cls, "tier": tier, "reason": "within policy"}
        if action == "redact":
            return {"action": "redact", "payload": redact(payload), "classification": cls, "tier": tier, "reason": f"{cls} redacted for {tier}"}
        return {"action": "block", "payload": None, "classification": cls, "tier": tier, "reason": f"{cls} not permitted to {tier}"}
    except Exception:  # noqa: BLE001 — fail-closed
        return {"action": "block", "payload": None, "classification": "error", "tier": str(destination_tier or ""), "reason": "egress guard error (fail-closed)"}


# Provider name → destination tier. Local providers never trigger redaction;
# everything else is an external third party.
_LOCAL_PROVIDERS = {"ollama", "local", "llama_cpp", "vllm_local"}


def tier_for_provider(provider: str) -> str:
    return "local" if str(provider or "").lower() in _LOCAL_PROVIDERS else "external_api"


def guard_for_provider(payload: Any, provider: str) -> dict[str, Any]:
    """Convenience: guard a payload for a named LLM provider."""
    return guard(payload, tier_for_provider(provider))
